mypy/waiter.py

```diff
@@ -239,22 +269,81 @@ def run(self) -> int:
         if self.verbosity == 0:
             self._note = Noter(len(self.queue))
         print('SUMMARY  %d tasks selected' % len(self.queue))
+
+        def avg(lst: Iterable[float]) -> float:
+            valid_items = [item for item in lst if item is not None]
+            if not valid_items:
+                # we don't know how long a new task takes
+                # better err by putting it in front in case it is slow:
+                # a fast task in front hurts performance less than a slow task in the back
+                return float('inf')
+            else:
+                return sum(valid_items) / len(valid_items)
+
+        logs = self.load_log_file()
+        if logs:
+            times = {cmd.name: avg(log['runtime'].get(cmd.name, None) for log in logs)
+                     for cmd in self.queue}
+
+            def sort_function(cmd: LazySubprocess) -> Tuple[Any, int, float]:
+                # longest tasks first
+                runtime = -times[cmd.name]
+                # sequential tasks go first by default
+                sequential = -(cmd.name in self.sequential_tasks)
+                if self.ff:
+                    # failed tasks first with -ff
+                    exit_code = -logs[-1]['exit_code'].get(cmd.name, 0)
+                    if not exit_code:
+                        # avoid interrupting parallel tasks with sequential in between
+                        # => order: seq failed, parallel failed, parallel passed, seq passed
+                        # => among failed tasks, sequential should go before parallel
+                        # => among successful tasks, sequential should go after parallel
+                        sequential = -sequential
```

Comment:
Editorial nit: what do the `=>` arrows signify? Did you mean an unordered list? Adding punctuation at the end of an element also improves reading flow. Compare:

> avoid interrupting parallel tasks with sequential in between
> * order: seq failed, parallel failed, parallel passed, seq passed;
> * among failed tasks, sequential should go before parallel; and
> * among successful tasks, sequential should go after parallel.